File: examples/pyspark-compute-pi/compute_pi.py

from random import random

from pyspark.sql import SparkSession

N_SLICE = 10
N_THROWS = 100000 * N_SLICE

def throwDarts(_):
    x = random() * 2 - 1
    y = random() * 2 - 1

    return 1 if x ** 2 + y ** 2 <= 1 else 0


def compute_pi(spark: SparkSession): 

  num_list = list(range(N_THROWS))

  incremental_rdd = spark.sparkContext.parallelize(num_list)

  # map 
  darts_rdd = incremental_rdd.map(throwDarts)

  # reduce
  darts_in_circle = darts_rdd.reduce(lambda a,b: a+b)

  print("Pi is roughly {}".format(4.0 * darts_in_circle / N_THROWS))



if __name__ == "__main__": 

  # start a session on local cluster
  spark = SparkSession.builder\
      .appName("pyspark-compute-pi")\
      .master("spark://localhost:7077")\
      .getOrCreate()

  # run a job
  compute_pi(spark)

  # terminate the session
  spark.stop()

File: examples/spark-compute-pi/src/main/scala/my/example/pi_compute/compute_pi.scala
package my.example.pi_compute
package my.example.pi_compute

import org.apache.spark.sql.SparkSession


// singleton, invoked at most once
object PiComputeScalaExample {


    def main(args: Array[String]): Unit = {

      val NUM_SAMPLES: Int = 100000

      val spark = SparkSession.builder
                    .appName("Spark Pi Compute")
                    .master("spark://localhost:7077")
                    .getOrCreate

      val count = spark.sparkContext.parallelize(1 to NUM_SAMPLES).filter {
        _ => 
          val x = math.random
          val y = math.random

          x * x + y * y < 1
      }.count()

      println(s"Pi is roughly ${4.0 * count / NUM_SAMPLES}")

    }
}